File: app/Module.java
import com.codahale.metrics.ConsoleReporter;
import com.codahale.metrics.MetricRegistry;
import com.codahale.metrics.Slf4jReporter;
import com.google.inject.AbstractModule;
import hbase.connector.HBaseConnector;
import hbase.connector.HBaseInstanceConnector;
import hbase.load.AdminDataLoad;
import hbase.load.HBaseAdminDataLoader;
import hbase.repository.AdminDataRepository;
import hbase.repository.HBaseAdminDataRepository;
import org.apache.commons.lang3.StringUtils;
import org.apache.hadoop.hbase.*;
import org.apache.hadoop.hbase.client.Connection;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import play.Configuration;
import play.Environment;

import javax.inject.Inject;
import javax.inject.Provider;
import java.io.IOException;
import java.util.concurrent.TimeUnit;


/**
 * This class is a Guice module that tells Guice how to bind several
 * different types. This Guice module is created when the Play
 * application starts.
 * <p>
 * Play will automatically use any class called `Module` that is in
 * the root package. You can create modules in other locations by
 * adding `play.modules.enabled` settings to the `application.conf`
 * configuration file.
 */
public class Module extends AbstractModule {

    private final Configuration configuration;

    public Module(Environment environment, Configuration configuration) {
        this.configuration = configuration;
    }

    @Override
    public void configure() {
        if (configuration.getBoolean("hbase.initialize")) {
            System.setProperty(RepositoryInitializer.HEADER_KEY, configuration.getString(RepositoryInitializer.HEADER_KEY));
            bind(RepositoryInitializer.class).asEagerSingleton();
        }
        bind(HBaseConnector.class).to(HBaseInstanceConnector.class);
        bind(AdminDataRepository.class).to(HBaseAdminDataRepository.class);
        bind(AdminDataLoad.class).to(HBaseAdminDataLoader.class);
        if (configuration.getBoolean("api.metrics")) {
            bind(MetricRegistry.class).toProvider(MetricRegistryProvider.class);
        }
    }
}

class RepositoryInitializer {

    public static final String HEADER_KEY = "csv.header.string";

    private static final Logger LOG = LoggerFactory.getLogger(RepositoryInitializer.class);

    private final HBaseConnector connector;

    private final Configuration configuration;

    @Inject
    public RepositoryInitializer(HBaseConnector connector, Configuration configuration, AdminDataLoad dataLoader) throws IOException {
        this.connector = connector;
        this.configuration = configuration;

        LOG.info("Started repository initialization. Will create namespace and table if necessary.");
        final String namespace = configuration.getString("hbase.namespace");
        if (StringUtils.isNotBlank(namespace) && (!namespaceExists(namespace))) createNamespace(namespace);
        final TableName tableName = TableName.valueOf(namespace, configuration.getString("hbase.table.name"));
        if (!tableExists(tableName)) createTable(tableName);

        LOG.info("Invoking load of admin data.");
        dataLoader.load(tableName.getNameWithNamespaceInclAsString(), "201706", configuration.getString("csv.file"), configuration.getInt("csv.id.position"), configuration.getString(HEADER_KEY));
    }

    private void createTable(TableName tableName) throws IOException {
        try(Connection connection = connector.getConnection()) {
            HTableDescriptor htable = new HTableDescriptor(tableName);
            htable.addFamily( new HColumnDescriptor(configuration.getString("hbase.column.family")));
            connection.getAdmin().createTable(htable);
            LOG.info("Created table {}", tableName);

        } catch (TableExistsException e) {
            LOG.warn("Failed to create table {} since it already exists", tableName);
            LOG.debug("Failed to create table " + tableName, e);

        } catch (IOException e) {
            LOG.error("Failed to create table " + tableName, e);
            throw e;
        }
    }

    private boolean tableExists(TableName tableName) throws IOException {
        try(Connection connection = connector.getConnection()) {
            return connection.getAdmin().tableExists(tableName);
        }
    }

    private void createNamespace(String namespace) throws IOException {
        try(Connection connection = connector.getConnection()) {
            connection.getAdmin().createNamespace(NamespaceDescriptor.create(namespace).build());
            LOG.info("Created namespace {}", namespace);

        } catch (NamespaceExistException e) {
            LOG.warn("Failed to create namespace {} since it already exists", namespace);
            LOG.debug("Failed to create namespace " + namespace, e);

        } catch (IOException e) {
            LOG.error("Failed to create namespace " + namespace, e);
            throw e;
        }
    }

    private boolean namespaceExists(String namespace) throws IOException {
        try(Connection connection = connector.getConnection()) {
            connection.getAdmin().getNamespaceDescriptor(namespace);
            return true;

        } catch (NamespaceNotFoundException e) {
            return false;
        }
    }
}


class MetricRegistryProvider implements Provider<MetricRegistry> {
    private static final Logger logger = LoggerFactory.getLogger("application.Metrics");

    private static final MetricRegistry registry = new MetricRegistry();

    private void consoleReporter() {
        ConsoleReporter reporter = ConsoleReporter.forRegistry(registry)
                .convertRatesTo(TimeUnit.SECONDS)
                .convertDurationsTo(TimeUnit.MILLISECONDS)
                .build();
        reporter.start(1, TimeUnit.SECONDS);
    }

    private void slf4jReporter() {
        final Slf4jReporter reporter = Slf4jReporter.forRegistry(registry)
                .outputTo(logger)
                .convertRatesTo(TimeUnit.SECONDS)
                .convertDurationsTo(TimeUnit.MILLISECONDS)
                .build();
        reporter.start(1, TimeUnit.MINUTES);
    }

    public MetricRegistryProvider() {
        consoleReporter();
        slf4jReporter();
    }

    @Override
    public MetricRegistry get() {
        return registry;
    }
}

File: app/Module.scala
import java.util.concurrent.TimeUnit
import javax.inject.Provider

import play.{Configuration, Environment}
import org.slf4j.LoggerFactory
import com.codahale.metrics.{ConsoleReporter, MetricRegistry, Slf4jReporter}
import com.google.inject.AbstractModule

import hbase.connector.{HBaseConnector, HBaseInMemoryConnector}
import hbase.load.{AdminDataLoad, BulkLoader, HBaseAdminDataLoader}
import hbase.repository.{AdminDataRepository, InMemoryAdminDataRepository, RestAdminDataRepository}

/**
  * Module
  * ----------------
  * Date: 01 February 2018 - 12:22
  * Copyright (c) 2017  Office for National Statistics
  */
class Module(val environment: Environment, val configuration: Configuration) extends AbstractModule {
  override def configure(): Unit = {
    if (configuration.getBoolean("hbase.initialize")) {
      System.setProperty(BulkLoader.HEADER_STRING, configuration.getString("csv.header.string"))
      bind(classOf[HBaseConnector]).toInstance(new HBaseInMemoryConnector(configuration.getString("hbase.table.name")))
      bind(classOf[RepositoryInitializer]).asEagerSingleton()
      bind(classOf[AdminDataRepository]).to(classOf[InMemoryAdminDataRepository]).asEagerSingleton()
      bind(classOf[AdminDataLoad]).to(classOf[HBaseAdminDataLoader]).asEagerSingleton()
    } else { bind(classOf[AdminDataRepository]).to(classOf[RestAdminDataRepository]).asEagerSingleton() }
    if (configuration.getBoolean("api.metrics")) {
      bind(classOf[MetricRegistry])
        .toProvider(classOf[MetricRegistryProvider])
        .asEagerSingleton()
    }
  }
}


class MetricRegistryProvider() extends Provider[MetricRegistry] {
  consoleReporter()
  slf4jReporter()

  private def consoleReporter(): Unit = {
    val reporter = ConsoleReporter
      .forRegistry(MetricRegistryProvider.REGISTRY)
      .convertRatesTo(TimeUnit.SECONDS)
      .convertDurationsTo(TimeUnit.MILLISECONDS)
      .build
    reporter.start(1, TimeUnit.SECONDS)
  }

  private def slf4jReporter(): Unit = {
    val reporter = Slf4jReporter
      .forRegistry(MetricRegistryProvider.REGISTRY)
      .outputTo(MetricRegistryProvider.LOGGER)
      .convertRatesTo(TimeUnit.SECONDS)
      .convertDurationsTo(TimeUnit.MILLISECONDS)
      .build
    reporter.start(1, TimeUnit.MINUTES)
  }

  override def get: MetricRegistry = MetricRegistryProvider.REGISTRY
}

object MetricRegistryProvider {
  private val LOGGER = LoggerFactory.getLogger("application.Metrics")
  private val REGISTRY = new MetricRegistry
}